### PR TITLE
AP_Compass: set only if changed to save space in param storage while reordering

### PR DESCRIPTION
**Pull request by bugobliterator**

### CODE REVIEW

**Comment by magicrub** [comment]
wouldn't we want to just replace set_and_save() with this logic?

File: libraries/AP_Param/AP_Param.h [hunk 1]
```diff
@@ -887,6 +887,28 @@ class AP_ParamV : public AP_Param
...
+    /// Combined set and save, but only does the save if the value is
+    /// different from the current ram value, thus saving us a
+    /// scan(). This should only be used where we have not set() the
+    /// value separately, as otherwise the value in EEPROM won't be
+    /// updated correctly.
+    void set_and_save_ifchanged(const T &v) {
+        if (vtype != AP_PARAM_VECTOR3F) {
+            // if not Vector3f type we just 
+            // set and save normally
+            set_and_save(v);
```
Comment (tridge) [inline libraries/AP_Param/AP_Param.h:899]: why should an ifchanged function call set_and_save()?

File: libraries/AP_Param/AP_Param.h [hunk 2]
```diff
@@ -887,6 +887,28 @@ class AP_ParamV : public AP_Param
         save(force);
     }
 
+    /// Combined set and save, but only does the save if the value is
+    /// different from the current ram value, thus saving us a
+    /// scan(). This should only be used where we have not set() the
+    /// value separately, as otherwise the value in EEPROM won't be
+    /// updated correctly.
+    void set_and_save_ifchanged(const T &v) {
+        if (vtype != AP_PARAM_VECTOR3F) {
```
Comment (tridge) [inline libraries/AP_Param/AP_Param.h:896]: this is inside the ParamV class, how can it not be a vector?

**Review by tridge — CHANGES_REQUESTED** [review]
What testing has been done?
The critical thing is that there must be no place where we call set() on these variables for set_and_save_ifchanged() to be effective. If we call set() to set a value without saving and then later call set_and_save_ifchanged() then the save won't happen and the values won't persist.

**Comment by tridge** [comment]
> wouldn't we want to just replace set_and_save() with this logic?

because we rely on forcing storage of values in many places. For example, we rely on it for being able to tell if IMUs and compasses are calibrated

**Comment by tridge** [comment]
I see quite a few places in AP_Compass where set() is used on these variables. It will take some careful checking and testing to make sure this change is safe

File: libraries/AP_Param/AP_Param.h (hunk 2 above)
Comment (bugobliterator) [reply libraries/AP_Param/AP_Param.h:896]: that's right will get rid of this.

**Comment by bugobliterator** [comment]
> I see quite a few places in AP_Compass where set() is used on these variables. It will take some careful checking and testing to make sure this change is safe

Yeah I already checked this, reorder compass is only ever called inside init method, and we don't do a single param set() only call before _reorder_compass_params() under init(). Only issue I can see is if in the future someone by mistake moves this method, or uses during run time. I can leave a comment not to do that.

**Comment by bugobliterator** [comment]
For testing, first it passes SITL reorder test written by Peter in previous PR. Also I tested it manually by doing calibration and moving values around and see if the values moved persistently. Also placed debug prints under the set_and_save_ifchanged command, and confirmed that it doesn't save params which are unchanged from default and saves those which have been changed.

**Review by tridge — APPROVED** [review]

**Pull request merged**